cmd/catchpointdump/file.go

```diff
@@ -0,0 +1,225 @@
+	}
+	if fileHeader.Version != 0 {
+		fmt.Fprintf(outFile, "Version: %d\nBalances Round: %d\nBlock Round: %d\nBlock Header Digest: %s\nCatchpoint: %s\nTotal Accounts: %d\nTotal Chunks: %d\n",
+			fileHeader.Version,
+			fileHeader.BalancesRound,
+			fileHeader.BlocksRound,
+			fileHeader.BlockHeaderDigest.String(),
+			fileHeader.Catchpoint,
+			fileHeader.TotalAccounts,
+			fileHeader.TotalChunks)
+
+		totals := fileHeader.Totals
+		fmt.Fprintf(outFile, "AccountTotals - Online Money: %d\nAccountTotals - Online RewardUnits : %d\nAccountTotals - Offline Money: %d\nAccountTotals - Offline RewardUnits : %d\nAccountTotals - Not Participating Money: %d\nAccountTotals - Not Participating Money RewardUnits: %d\nAccountTotals - Rewards Level: %d\n",
+			totals.Online.Money.Raw, totals.Online.RewardUnits,
+			totals.Offline.Money.Raw, totals.Offline.RewardUnits,
+			totals.NotParticipating.Money.Raw, totals.NotParticipating.RewardUnits,
+			totals.RewardsLevel)
+	}
+	return dbAccessor.Atomic(func(tx *sql.Tx) (err error) {
+		if fileHeader.Version == 0 {
+			var totals ledger.AccountTotals
+			id := ""
+			row := tx.QueryRow("SELECT online, onlinerewardunits, offline, offlinerewardunits, notparticipating, notparticipatingrewardunits, rewardslevel FROM accounttotals WHERE id=?", id)
+			err = row.Scan(&totals.Online.Money.Raw, &totals.Online.RewardUnits,
+				&totals.Offline.Money.Raw, &totals.Offline.RewardUnits,
+				&totals.NotParticipating.Money.Raw, &totals.NotParticipating.RewardUnits,
+				&totals.RewardsLevel)
+			if err != nil {
+				return err
+			}
+			fmt.Fprintf(outFile, "AccountTotals - Online Money: %d\nAccountTotals - Online RewardUnits : %d\nAccountTotals - Offline Money: %d\nAccountTotals - Offline RewardUnits : %d\nAccountTotals - Not Participating Money: %d\nAccountTotals - Not Participating Money RewardUnits: %d\nAccountTotals - Rewards Level: %d\n",
+				totals.Online.Money.Raw, totals.Online.RewardUnits,
+				totals.Offline.Money.Raw, totals.Offline.RewardUnits,
+				totals.NotParticipating.Money.Raw, totals.NotParticipating.RewardUnits,
+				totals.RewardsLevel)
+		}
+
+		balancesTable := "accountbase"
+		if fileHeader.Version != 0 {
+			balancesTable = "catchpointbalances"
+		}
+		rows, err := tx.Query(fmt.Sprintf("SELECT address, data FROM %s order by address", balancesTable))
+		if err != nil {
+			return
+		}
+		defer rows.Close()
+
+		for rows.Next() {
+			var addrbuf []byte
+			var buf []byte
+			err = rows.Scan(&addrbuf, &buf)
+			if err != nil {
+				return
+			}
+
+			var data basics.AccountData
+			err = protocol.Decode(buf, &data)
+			if err != nil {
+				return
+			}
+
+			var addr basics.Address
+			if len(addrbuf) != len(addr) {
+				err = fmt.Errorf("Account DB address length mismatch: %d != %d", len(addrbuf), len(addr))
+				return
+			}
+			copy(addr[:], addrbuf)
+			jsonData, err := json.Marshal(data)
+			if err != nil {
+				return err
+			}
+
+			fmt.Fprintf(outFile, "%v : %s\n", addr, string(jsonData))
```

Comment:
If outFile is blocked, the database is blocked. 
If this tool is used on a production database, and accidentally the output is blocked (e.g. printing to screen with paging), it can be a problem. 
Maybe put a timeout?

Reply:
I don't expect to be using it directly on production servers.. the intention here is to allow offline database to be examined. Also, when dumping a catchpoint, it creates it's own database, so there are no locking contentions.